### PR TITLE
Rollup of changes

### PR DESCRIPTION
**Pull request by nephele-gh**
e18ef0fd85691b817158ba532ab5454e27b13fd1 requires https://review.haiku-os.org/c/haiku/+/4571 to be merged first

### CODE REVIEW

File: Source/WebCore/html/ImageDocument.cpp [hunk 1]
```diff
@@ -229,6 +230,13 @@ void ImageDocument::createDocumentStructure()
 
     auto body = HTMLBodyElement::create(*this);
     body->setAttribute(styleAttr, "margin: 0px");
+#if PLATFORM(HAIKU)
```
Comment (waddlesplash) [inline Source/WebCore/html/ImageDocument.cpp:233]: This seems like the wrong way to do this. Isn't there a general "document colors" (aka, HTML3 colors) that we should be able to set at a higher level?

Comment (nephele-gh) [reply Source/WebCore/html/ImageDocument.cpp:233]: Yes, or there should be.
In the other patch i have these two caces for css color variables CSSValueCanvas and CSSValueWindow. IIRC the gtk port also uses these two cases, but I have not gotten it to work correctly for a default canvas color. It is possible these are only set for some "this is the default color" css variables from css color module level 4
(https://www.w3.org/TR/css-color-4/#css-system-colors)
There a Canvas color name is defined.

I could probably try setting the background to Canvas and see if that helps.
(on another related note :root { color-scheme: light dark; } should be enough to change the document background color, but it does nothing with the color patch... I don't know why, the other css values seem to work)

**Comment by nephele-gh** [comment]
This should be good to go now.
One minor issue is that button border color is not exposed, even though we have this. Webkit seems to be missing a var here.

And for another thing the default canvas and default link color etc don't seem to be used in the default html document :D.
But that can be tackled later.

As an additional thing the Haiku: support rendering dark/light controls patch fixes a todo with clicking state and exposes the hovered state of webkit buttons to the controllook. Which means now buttons in native drawing respond to hovering, hooray.

**Comment by nephele-gh** [comment]
-1
I want to rework this a bit, the colorForValue call should be shared instead. 

The commit currently will also pick the wrong scrollbar thumb color.

**Comment by nephele-gh** [comment]
Okay, the review is ready.
I need to test it once more with my haiku commit merged now, but it should be good to go!

File: Source/WebCore/css/html.css [hunk 2]
```diff
@@ -1272,13 +1277,13 @@ input[type="file"]:focus::file-selector-button {
 }
 
 a:any-link {
-    color: -webkit-link;
+	color: Linktext;
```
Comment (pulkomandy) [inline Source/WebCore/css/html.css:1280]: looks like a mix of tabs and space here

File: Source/WebCore/css/html.css [hunk 3]
```diff
@@ -23,6 +23,11 @@
 
 @namespace "http://www.w3.org/1999/xhtml";
 
+:root {
```
Comment (pulkomandy) [inline Source/WebCore/css/html.css:26]: this is modifying a file shared with other platforms, has it been discussed with upstream on webkit bugtracker? To make sure we are doing this correctly

File: Source/WebCore/platform/haiku/RenderThemeHaiku.h [hunk 4]
```diff
@@ -91,8 +91,13 @@ class RenderThemeHaiku : public RenderTheme {
 
     bool paintSliderThumb(const RenderObject&, const PaintInfo&, const IntRect&) override;
 
+    Color systemColor(CSSValueID, OptionSet<StyleColorOptions>) const;
```
Comment (pulkomandy) [inline Source/WebCore/platform/haiku/RenderThemeHaiku.h:94]: please add the "override" keyword for functions that overrides the parent class

File: Source/WebCore/css/html.css (hunk 3 above)
Comment (pulkomandy) [reply Source/WebCore/css/html.css:26]: Generally for me it's better if all changes to common files are upstreamed, then I have no conflicts when merging changes in our branch. It's not a reason to block this change, but if the plan is to have this as a patch here forever, I prefer to know about it. Or maybe if WebKi devs can explain how they do it in other platforms and we can see if it's possible to reuse the same way they do?

**Comment by nephele-gh** [comment]
I've talked to upstream, the change in Source/WebCore/html/ImageDocument.cpp I will submit to webkit directly (they also had a similar change for text documents recently, which is nice!)

The document color can be set in FrameView::setBaseBackgroundColor, I will investigate that tommorow, additionally upstream also has a pr to go from text->CanvasText, but it should likely be sufficient to add the css value for "text" to the system color thing (https://bugs.webkit.org/show_bug.cgi?id=232954)

Link color might be a bit more finicky, upstream also has issues with it
(https://bugs.webkit.org/show_bug.cgi?id=209851)

**Comment by nephele-gh** [comment]
https://github.com/WebKit/WebKit/pull/2734

**Comment by nephele-gh** [comment]
Another equivalent change for CanvasText instead of text was merged in webkit, I would redo this with the next versioing commit, or alternatively prepare a PR for the next versioing commit, and then  rebase the LR on that.

(currently there is a bit of an issue to get the image one merged, seems I brole the webkit tool, but I'll resubmit it via their bugzilla next werk if the tool still fails then)

**Comment by pulkomandy** [comment]
I can merge the WebKit upstream without waiting for an upstream "versioning" commit if that helps (and if the intermediate commit actually builds and runs, often they are broken if I just pick a random one).

**Comment by nephele-gh** [comment]
Okay, that's good to know!
I think this https://github.com/WebKit/WebKit/pull/2706 still needs to be merged and my images commit, I'll let you know once that is done.

File: Source/WebCore/platform/haiku/RenderThemeHaiku.h (hunk 4 above)
Comment (nephele-gh) [reply Source/WebCore/platform/haiku/RenderThemeHaiku.h]: done

**Comment by nephele-gh** [comment]
Should be good now. This commit doesn't fix the background color of the default html though. I will do that in a later commit if i can find a good way to do it.

**Comment by nephele-gh** [comment]
> This commit doesn't fix the background color of the default html though. I will do that in a later commit if i can find a good way to do it.

Fixed! It now does this aswell :)

**Pull request merged**